pkg/e2e/infra/utils.go

```diff
@@ -422,3 +424,138 @@ func waitUntilNodesAreDeleted(client runtimeclient.Client, listOpt *runtimeclien
+			return false, nil
+		}
+		return true, nil
+	})
+}
+
+func verifyNodeDraining(client runtimeclient.Client, targetMachine *mapiv1beta1.Machine, rc *corev1.ReplicationController) (string, error) {
+	var drainedNodeName string
+	err := wait.PollImmediate(e2e.RetryMedium, e2e.WaitLong, func() (bool, error) {
+		machine := mapiv1beta1.Machine{}
+
+		key := types.NamespacedName{
+			Namespace: targetMachine.Namespace,
+			Name:      targetMachine.Name,
+		}
+		if err := client.Get(context.TODO(), key, &machine); err != nil {
+			glog.Errorf("Error querying api machine %q object: %v, retrying...", targetMachine.Name, err)
+			return false, nil
+		}
+		if machine.Status.NodeRef == nil || machine.Status.NodeRef.Kind != "Node" {
+			glog.Error("Machine %q not linked to a node", machine.Name)
+		}
+
+		drainedNodeName = machine.Status.NodeRef.Name
+		node := corev1.Node{}
+
+		if err := client.Get(context.TODO(), types.NamespacedName{Name: drainedNodeName}, &node); err != nil {
+			glog.Errorf("Error querying api node %q object: %v, retrying...", drainedNodeName, err)
+			return false, nil
+		}
+
+		if !node.Spec.Unschedulable {
+			glog.Errorf("Node %q is expected to be marked as unschedulable, it is not", node.Name)
+			return false, nil
+		}
+
+		glog.Infof("Node %q is mark unschedulable as expected", node.Name)
+
+		pods := corev1.PodList{}
+		listOpt := &runtimeclient.ListOptions{}
+		listOpt.MatchingLabels(rc.Spec.Selector)
+		if err := client.List(context.TODO(), listOpt, &pods); err != nil {
+			glog.Errorf("Error querying api for Pods object: %v, retrying...", err)
+			return false, nil
+		}
+
+		// expecting nodeGroupSize nodes
+		podCounter := 0
+		for _, pod := range pods.Items {
+			if pod.Spec.NodeName != machine.Status.NodeRef.Name {
+				continue
+			}
+			if !pod.DeletionTimestamp.IsZero() {
+				continue
+			}
+			podCounter++
+		}
+
+		glog.Infof("Have %v pods scheduled to node %q", podCounter, machine.Status.NodeRef.Name)
+
+		// Verify we have enough pods running as well
+		rcObj := corev1.ReplicationController{}
+		key = types.NamespacedName{
+			Namespace: rc.Namespace,
+			Name:      rc.Name,
+		}
+		if err := client.Get(context.TODO(), key, &rcObj); err != nil {
+			glog.Errorf("Error querying api RC %q object: %v, retrying...", rc.Name, err)
+			return false, nil
+		}
+
+		// The point of the test is to make sure majority of the pods is rescheduled
+		// to other nodes. Pod disruption budget makes sure at most one pod
+		// owned by the RC is not Ready. So no need to test it. Though, usefull to have it printed.
+		glog.Infof("RC ReadyReplicas/Replicas: %v/%v", rcObj.Status.ReadyReplicas, rcObj.Status.Replicas)
+
+		// This makes sure at most one replica is not ready
+		if rcObj.Status.Replicas-rcObj.Status.ReadyReplicas > 1 {
+			return false, fmt.Errorf("pod disruption budget not respected, node was not properly drained")
+		}
+
+		// Depends on timing though a machine can be deleted even before there is only
```

Comment:
If it is dependent on timing, is this teat reliable/repeatable and/or still deterministic?

Reply:
It's still reliable. The thing is a machine is delete right after the drain operation succeeds. And the drain operation succeeds when there is no pod running on a node. So in case the operation is quick (e.g. it's easy to delete pod that is running just `sleep` command), the retry period of `e2e.RetryMedium` (5 seconds) can be too long to capture a case where there is only one running pod left. One could argue even those two pods could be deleted in the interval. Thus the reason why there is 10 pods to be evicted so the chance is minimized.